zkevm-circuits/src/evm_circuit/execution/codecopy.rs

```diff
@@ -207,217 +207,32 @@ impl<F: Field> ExecutionGadget<F> for CodeCopyGadget<F> {
-    };
-
-    fn test_ok(src_addr: u64, dst_addr: u64, size: usize) {
-        let randomness = Fr::rand();
-        let call_id = 1;
+    use crate::test_util::run_test_circuits;
 
+    fn test_ok(memory_offset: usize, code_offset: usize, size: usize) {
         let code = bytecode! {
             PUSH32(Word::from(size))
-            PUSH32(Word::from(src_addr))
-            PUSH32(Word::from(dst_addr))
+            PUSH32(Word::from(code_offset))
+            PUSH32(Word::from(memory_offset))
             CODECOPY
             STOP
         };
-        let code = Bytecode::new(code.to_vec());
-
-        let mut rws_map = RwMap(
-            [(
-                RwTableTag::Stack,
-                vec![
-                    // Stack item written by PUSH32.
-                    Rw::Stack {
-                        rw_counter: 1,
-                        is_write: true,
-                        call_id,
-                        stack_pointer: 1023,
-                        value: Word::from(size),
-                    },
-                    // Stack item written by PUSH32.
-                    Rw::Stack {
-                        rw_counter: 2,
-                        is_write: true,
-                        call_id,
-                        stack_pointer: 1022,
-                        value: Word::from(src_addr),
-                    },
-                    // Stack item written by PUSH32.
-                    Rw::Stack {
-                        rw_counter: 3,
-                        is_write: true,
-                        call_id,
-                        stack_pointer: 1021,
-                        value: Word::from(dst_addr),
-                    },
-                    // First stack item read by CODECOPY.
-                    Rw::Stack {
-                        rw_counter: 4,
-                        is_write: false,
-                        call_id,
-                        stack_pointer: 1021,
-                        value: Word::from(dst_addr),
-                    },
-                    // Second stack item read by CODECOPY.
-                    Rw::Stack {
-                        rw_counter: 5,
-                        is_write: false,
-                        call_id,
-                        stack_pointer: 1022,
-                        value: Word::from(src_addr),
-                    },
-                    // Third stack item read by CODECOPY.
-                    Rw::Stack {
-                        rw_counter: 6,
-                        is_write: false,
-                        call_id,
-                        stack_pointer: 1023,
-                        value: Word::from(size),
-                    },
-                ],
-            )]
-            .into(),
+        assert_eq!(
+            run_test_circuits(
+                TestContext::<2, 1>::simple_ctx_with_bytecode(code).unwrap(),
+                None,
+            ),
+            Ok(()),
```

Comment:
As far as negative tests are concerned, we can have insufficient gas or stack underflow. But since the witness is generated from the bus-mapping, I will need to dig deeper into how I can force a stack underflow. May be this can be tracked as a separate issue, since those negative tests will need to be implemented for all other opcodes as well. Wdyt?